components/api/src/main/java/com/hotels/styx/api/extension/service/spi/StyxService.java

```diff
@@ -73,4 +76,35 @@
     default Map<String, HttpHandler> adminInterfaceHandlers() {
         return emptyMap();
     }
+
+    /**
+     * Derives a new service interface with added side-effects for errors.
+     * This could be used for logging, metrics, etc.
+     *
+     * @param consumer error consumer
+     * @return a new service interface
+     */
+    default StyxService doOnError(Consumer<Throwable> consumer) {
+        StyxService parent = this;
+
+        return new StyxService() {
+            @Override
+            public CompletableFuture<Void> start() {
+                return parent.start().exceptionally(throwable -> {
+                    consumer.accept(throwable);
+                    throw propagate(throwable);
+                });
+            }
+
+            @Override
+            public CompletableFuture<Void> stop() {
+                return parent.stop();
+            }
+
+            @Override
+            public Map<String, HttpHandler> adminInterfaceHandlers() {
+                return parent.adminInterfaceHandlers();
+            }
+        };
+    }
```

Comment:
This change seems stem from the synchronous server start in `ServerService` class. I think this change can be avoided.

If I understand correctly, we have two or three requirements:
1. Set the server state to "complete" when it starts successfully
2. Set the server state to "failed" when it fails to start.

and optionally 

3. Set the server state to "shut" etc when it is stopped 

You can achieve these by deriving `ServerService` directly from `AbstractStyxService` and overriding its `startService` method to start the HTTP server. You can then register a listener to get notified of the server state changes. Failing that, you could perhaps start the HTTP server in `CompletableFuture.supplyAsync` block and use `thenAccept` and `exceptionally` to capture the outcome.

This should knock off lots of code in `StyxService` and `BaseService` classes.